src/blocks/service_status.rs

```diff
@@ -0,0 +1,143 @@
+//! Display the status of a service
+//!
+//! Right now only `systemd` is supported.
+//!
+//! # Configuration
+//!
+//! Key | Values | Default
+//! ----|--------|--------
+//! `driver` | Which init system is running the service. Available drivers are: `"systemd"` | `"systemd"`
+//! `service` | The name of the service | **Required**
+//! `format` | A string to customise the output of this block. See below for available placeholders. | `" $service $active_state "`
+//!
+//! Placeholder    | Value                     | Type   | Unit
+//! ---------------|---------------------------|--------|-----
+//! `service`      | The name of the service   | Text   | -
+//! `active_state` | The status of the service | Text   | -
+//!
+//! # Example
+//!
+//! ```toml
+//! [[block]]
+//! block = "service_status"
+//! service = "cups"
+//! ```
+//!
+
+use super::prelude::*;
+use zbus::dbus_proxy;
+use zbus::PropertyStream;
+
+#[derive(Deserialize, Debug, Default)]
+#[serde(default)]
+pub struct Config {
+    driver: DriverType,
+    service: String,
+    format: FormatConfig,
+}
+
+#[derive(Deserialize, Debug, SmartDefault)]
+#[serde(rename_all = "snake_case")]
+enum DriverType {
+    #[default]
+    Systemd,
+}
+
+pub async fn run(config: Config, api: CommonApi) -> Result<()> {
+    let mut widget =
+        Widget::new().with_format(config.format.with_default(" $service $active_state ")?);
+
+    let mut driver: Box<dyn Driver> = match config.driver {
+        DriverType::Systemd => Box::new(SystemdDriver::new(config.service.clone()).await?),
+    };
+
+    loop {
+        let active_state = driver.is_active().await?;
+
+        let active_state_str = if active_state {
+            widget.state = State::Idle;
+            "active"
+        } else {
+            widget.state = State::Critical;
+            "inactive"
+        };
+
+        widget.set_values(map!(
+            "service" =>Value::text(config.service.clone()),
+            "active_state" =>Value::text(active_state_str.to_string()),
+        ));
+        api.set_widget(&widget).await?;
+
+        select! {
+            x = driver.wait_for_change() => x?,
+        }
```

Comment:
You might want to either call `api.event()` and ignore it, or close the `api.event_receiver` channel, otherwise the bar will deadlock eventually.